From: "Vishal Moola (Oracle)" <vishal.moola@gmail.com>
Subject: collapse_pte_mapped_thp() folio conversion fix
Date: Wed, 18 Oct 2023 14:24:47 -0700

filemap_lock_folio() can return an ERR_PTR on failure.
find_lock_page()/pagecache_get_page() handles this internally, while
filemap_lock_folio() does not. Ensure this is checked for in the caller
after converting find_lock_page() to filemap_lock_folio().

Link: https://lkml.kernel.org/r/ZTBjv4jyQ5/zu5cw@fedora
Signed-off-by: Vishal Moola (Oracle) <vishal.moola@gmail.com>
Reported-by: syzbot+1e2648076cadf48ad9a1@syzkaller.appspotmail.com
Closes: https://lkml.kernel.org/r/ZTBjv4jyQ5/zu5cw@fedora
Tested-by: syzbot+1e2648076cadf48ad9a1@syzkaller.appspotmail.com
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/khugepaged.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/mm/khugepaged.c~mm-khugepaged-convert-collapse_pte_mapped_thp-to-use-folios-fix
+++ a/mm/khugepaged.c
@@ -1506,7 +1506,7 @@ int collapse_pte_mapped_thp(struct mm_st
 
 	folio = filemap_lock_folio(vma->vm_file->f_mapping,
 			       linear_page_index(vma, haddr));
-	if (!folio)
+	if (!folio || IS_ERR(folio))
 		return SCAN_PAGE_NULL;
 
 	if (folio_order(folio) != HPAGE_PMD_ORDER) {
_
